### PR TITLE
[libbeat] : add processors "decode_duration", "move_fields"

### PR DESCRIPTION
**Pull request by kunsonx**
## What does this PR do?
Ability to manipulate log fields.

1. Give a prefix for all log fields
2. Decode go style duration string then custom format
3. Move fields from map into another map
4. Move fields into a parent field map.

## Why is it important?

Give users ability custom log field operations.

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have made corresponding change to the default configuration files
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have added an entry in `CHANGELOG.next.asciidoc` or `CHANGELOG-developer.next.asciidoc`.

## Author's Checklist

- [ ]

## Related issues

-

### CODE REVIEW

**Review by belimawr — COMMENTED** [review]
@kunsonx Are you still working on this PR? On a quick glance it looks fine, but it's missing tests and documentation.

Could you request a re-review once you're done with coding?

**Comment by kunsonx** [comment]
> @kunsonx Are you still working on this PR? On a quick glance it looks fine, but it's missing tests and documentation.
> 
> Could you request a re-review once you're done with coding?

OK. I will finish this works. Should I close this MR and open again when it's done?

**Comment by belimawr** [comment]
> > @kunsonx Are you still working on this PR? On a quick glance it looks fine, but it's missing tests and documentation.
> > Could you request a re-review once you're done with coding?
> 
> OK. I will finish this works. Should I close this MR and open again when it's done?
> 
> @belimawr
> 
> BTW Where is the guideline for write tests and documentation?
> 
> I am beginner for this project contribute.

No need to close/open another PR, just make sure to use the re-review request button once you're done so I get notified and can start reviewing. 

You can check the contributing docs here: https://www.elastic.co/guide/en/beats/devguide/current/beats-contributing.html, there you will also find information regarding tests and documentation.

If you still need something more specific, then try looking how the other processors implement their tests and documentation.

**Comment by kunsonx** [comment]
@belimawr Hi. I finished the `add_prefix` processor tests and documentation. Could you take a look it is satisfaction? if you feel it is ok then will finish these other processors.

Thanks for your time.

**Comment by belimawr** [comment]
> @belimawr Hi. I finished the `add_prefix` processor tests and documentation. Could you take a look it is satisfaction? if you feel it is ok then will finish these other processors.
> 
> Thanks for your time.

Hey @kunsonx I can already see some linters are failing, could you take a look at them? They're quite strict, so don't be dismayed by it. Usually it takes some iterations until all linter passes.

To run it locally, go the root of the repo and run
```
mage linter:lastChange
```

You will have to fix all lint issues in the files you changed (not only the lines you added).

Because we added this linter recently, sometimes we have to ask the linter to ignore some lines/functions with a `//nolint` directive.

**Comment by belimawr** [comment]
I'm closing it because there hasn't been any updated in a while.

@kunsonx, if you still want to work on this, just re-open the PR and push the new commits ;)

**Comment by kunsonx** [comment]
@belimawr Hi belimawr.

Just busy for while and rework of these processors.

Just take a look it is satisfy?

https://github.com/elastic/beats/pull/33119

Thanks.

**Comment by kunsonx** [comment]
Features sync from https://github.com/elastic/beats/pull/33119. please take a look. 

Thanks. @belimawr

File: libbeat/processors/decode_duration/decode_duration.go
Comment (belimawr) [inline libbeat/processors/decode_duration/decode_duration.go:73]: I know it's is a unlikely error to happen, but it's better to check and handle it.

File: libbeat/processors/move_fields/docs/move_fields.asciidoc
Comment (belimawr) [inline libbeat/processors/move_fields/docs/move_fields.asciidoc:68]: Small improvement on the copy:
```suggestion
To add a prefix to the whole event:
```

File: libbeat/processors/move_fields/docs/move_fields.asciidoc
Comment (belimawr) [inline libbeat/processors/move_fields/docs/move_fields.asciidoc:8]: We use the word "event" to refer to the data a Beat is ingesting/processing. Some other bits of this documentation also need this updating from "message" to "event".
```suggestion
The `move-fields` processor allow you to rearrange your event fields from a object into another object.
```

File: libbeat/processors/move_fields/docs/move_fields.asciidoc
Comment (belimawr) [inline libbeat/processors/move_fields/docs/move_fields.asciidoc:77]:
```suggestion
use this configuration
```

File: libbeat/processors/move_fields/docs/move_fields.asciidoc
Comment (belimawr) [inline libbeat/processors/move_fields/docs/move_fields.asciidoc:85]:
```suggestion
then your final event will be:
```

File: libbeat/processors/move_fields/docs/move_fields.asciidoc
Comment (belimawr) [inline libbeat/processors/move_fields/docs/move_fields.asciidoc:103]: If no configuration is required, what does the default config do?

It's better to have some of them required to make sure the processor will always do something useful. Based on your examples, it seems that at least `to` should be required.

File: libbeat/processors/move_fields/move_fields.go
Comment (belimawr) [inline libbeat/processors/move_fields/move_fields.go:66]: Here it seems you're checking if `parentValue` is an object, so the error message can be more direct, something like this:
```suggestion
		parent, ok = parentValue.(mapstr.M)
		if !ok {
			return nil, fmt.Errorf("'%s' does not contain an object, it contains '%v'", parentValue, p)
		}
```

File: libbeat/processors/move_fields/move_fields.go
Comment (belimawr) [inline libbeat/processors/move_fields/move_fields.go:60]: Here we can also improve a bit the wording:
```suggestion
			return nil, fmt.Errorf("cannot get value from key '%s": %w", p, err)
```

File: libbeat/processors/move_fields/move_fields.go
Comment (belimawr) [inline libbeat/processors/move_fields/move_fields.go:115]: Because you only want to know if the key is present on the map, you can use a `map[string]struct{}`. An empty struct is occupies zero bytes.
```suggestion
	fc.excludeMap = make(map[string]struct{})
	for _, k := range fc.Exclude {
		fc.excludeMap[k] = struct{}{}
	}
```

The rest of your code will still work because you're checking whether the keys is on the map. But the tests will need updating.

File: libbeat/processors/move_fields/move_fields_test.go
Comment (belimawr) [inline libbeat/processors/move_fields/move_fields_test.go:30]: Did you mean "expect" (like the expected result) here?

File: libbeat/processors/move_fields/move_fields_test.go
Comment (belimawr) [inline libbeat/processors/move_fields/move_fields_test.go:28]: Like the other test you can use [subtests](https://go.dev/blog/subtests) to create table-driven tests.

File: libbeat/processors/decode_duration/decode_duration_test.go
Comment (belimawr) [inline libbeat/processors/decode_duration/decode_duration_test.go:36]: You don't need this function here, you can just instantiate a `decodeDuration` whenever you need it, it's just a few lines. It will make it easier to the reader of the test to know the configuration of the `decodeDuration` processor.

**Review by belimawr — CHANGES_REQUESTED** [review]
Great Job!

I added some comments regarding the implementation and some documentation copy. I'll also try to get another reviewer at least for the docs.

**Comment by kunsonx** [comment]
> Great Job!
> 
> I added some comments regarding the implementation and some documentation copy. I'll also try to get another reviewer at least for the docs.

Hi. Thanks for your suggestion for this PR. 

All comments have already changed. 

Please take a look.

**Comment by belimawr** [comment]
run elasticsearch-ci/docs

File: libbeat/processors/move_fields/move_fields.go
Comment (belimawr) [inline libbeat/processors/move_fields/move_fields.go:52]: Aside the changes I mentioned on the documentation, I noticed another improvement we can make:
 - `excludeMap` is used at runtime by `moveFields`, it's not really a configuration field, so we can move it to `moveFields`.
 - I also ordered the fields by type and alphabetically.

```suggestion
type moveFieldsConfig struct {
	Exclude        []string `config:"exclude"`
	Fields         []string `config:"fields"`
	From           string   `config:"from"`
	To             string   `config:"to"`
	IgnoreMissing bool     `config:"ignore_missing"`
}

type moveFields struct {
	config     moveFieldsConfig
	excludeMap map[string]struct{}
}
```

File: libbeat/processors/move_fields/move_fields_test.go
Comment (belimawr) [inline libbeat/processors/move_fields/move_fields_test.go:44]: Using the input and expected result as name for the tests here does now work very well, so it's better to add a `name` field to your test case and use it when running the test. The name can be a short description of what you're testing on that case:
```suggestion
	cases := []struct {
		name         string
		in, expected mapstr.M
		p            *moveFields
	}{
		{
			"move fields to a new object",
			mapstr.M{"app": mapstr.M{"version": 1, "method": "2"}, "other": 3},
			mapstr.M{"app": mapstr.M{"method": "2"}, "rpc": mapstr.M{"version": 1}, "other": 3},
			&moveFields{config: moveFieldsConfig{
				ParentPath: "app",
				From:       nil,
				To:         "rpc.",
				Exclude:    []string{"method"},
				excludeMap: map[string]struct{}{"method": {}},
			}},
		},
```

**Review by belimawr — CHANGES_REQUESTED** [review]
Another set of comments, with some config names improvements.

Also your PR misses  adding add the processor docs in the processors list, to simplify I made a patch: [add_processors_to_list.patch.gz](https://github.com/elastic/beats/files/9753663/add_processors_to_list.patch.gz). If you have problems applying the patch, I can try to push the commit directly to your branch.

File: libbeat/processors/move_fields/move_fields.go
Comment (kunsonx) [reply libbeat/processors/move_fields/move_fields.go]: We can't put dot `.` here because the processor also supports adding a prefix to fields.

Comment (kunsonx) [reply libbeat/processors/move_fields/move_fields.go]: Sorry for the wrong click-to-commit suggestion.😂

**Comment by kunsonx** [comment]
The comment set has already changed except for one comment.

**Comment by belimawr** [comment]
@kunsonx, it looks great now! I'm just gonna get someone else to review the docs. In the mean time, could you rebase your PR on `main` (you can use the "Update Branch" option from GitHub)?

@dedemorton could you look at the doc files from this PR? I've already suggested some changes (that are already implemented), but it could use an expert set of eyes.

**Review by dedemorton — COMMENTED** [review]
Nice work getting everything tagged correctly and building! I have some edits that will hopefully make the content flow better. If the edits are accurate and you decide to accept them, I suggest batching all the commits into one (there's a button for that) before committing them in a single commit.

I would like to take a look again after the changes are merged. The diff is not the best way to see how well the changes work. Thanks!

File: libbeat/processors/move_fields/docs/move_fields.asciidoc
Comment (kunsonx) [reply libbeat/processors/move_fields/docs/move_fields.asciidoc:8]: I think "moves and optionally rearranges." maybe better.
The `move_fields` also can use for add prefix to fields. How we description it?

File: libbeat/processors/move_fields/docs/move_fields.asciidoc
Comment (kunsonx) [reply libbeat/processors/move_fields/docs/move_fields.asciidoc:90]: Yes. it is.

File: libbeat/processors/move_fields/docs/move_fields.asciidoc
Comment (kunsonx) [reply libbeat/processors/move_fields/docs/move_fields.asciidoc:91]: The `move_fields` processor first extract sub fields from `from` object then combine `to` will be new path.
For example the origin event is:
```json
{
  "a": {
    "name": "Jon",
    "age": 22
  }
}
```

use `move_fields` plugin set `from` = `a`, `to` = `my_prefix_` then the output is:
```json
{
  "a": {},
  "my_prefix_name": "Jon",
  "my_prefix_age": 22
}
```

I know it is hard to understand. so how should we to description it?

Comment (kunsonx) [reply libbeat/processors/move_fields/docs/move_fields.asciidoc:91]: Another example it is:
set `from` = `a`, `to` = 'b.' then the output is:
```json
{
  "a": {},
  "b": {
    "name": "Jon",
    "age": 22
  }
}
```

**Comment by kunsonx** [comment]
@dedemorton Thanks for your suggestions it make the docs better. Still has two suggestion here need more helps. 😀

**Comment by belimawr** [comment]
@dedemorton could you take another look/help out here?

File: libbeat/processors/move_fields/docs/move_fields.asciidoc
Comment (dedemorton) [reply libbeat/processors/move_fields/docs/move_fields.asciidoc:8]: Edited my suggestion. Feel free to commit if you like it!

File: libbeat/processors/move_fields/docs/move_fields.asciidoc
Comment (bmorelli25) [reply libbeat/processors/move_fields/docs/move_fields.asciidoc:91]: Is this any closer?

`from`: The fields to extract. This field and any nested fields will be moved to the `to` field unless they are filtered out. If `from` is empty, the event root is used. 

`to`: The destination of the extracted fields in the fields root. Added as a prefix to the fields extracted from `from`.

Something that's not clear to me in the two examples is why when `to` = `my_prefix_`, `my_prefix_` is prepended to the key values, but when `to` = `b`, the `name` and `age` fields are added as keys to the `b` object.

Comment (dedemorton) [reply libbeat/processors/move_fields/docs/move_fields.asciidoc:91]: Thanks Brandon! I think the syntax is a little confusing. The period in `b.` means that the fields get added to the `b` object. That makes me wonder if `to = b.my_prefix_` would also work and what the result would be. I am struggling to come up with a good suggestion here.

Comment (kunsonx) [reply libbeat/processors/move_fields/docs/move_fields.asciidoc:91]: Yes. it's also work.
Another example below:

config:
set `form` = ``, `fields` = `other`, `to` = `app.b.my_prefix_`, `exclude` = ``.

input event: 

```json
{
  "app": {
   "version": 1,
   "method": "2"
  },
  "other": 3
}
```

process output event: 

```json
{
  "app": {
   "version": 1,
   "method": "2"
   "b": {
       "my_prefix_other": 3
    }
  }
}
```

The processor can move fields to anywhere.

It any idea for the documentation?

Thanks. 😉

**Comment by kunsonx** [comment]
@dedemorton Have any idea about the pull request?

**Review by dedemorton — APPROVED** [review]
The changes LGTM!

**Review by belimawr — APPROVED** [review]

**Comment by belimawr** [comment]
@kunsonx could you update the PR with `main`? You can use the `Update branch` button here or do it manually.

There were some big changes in the codebase recently, but they should not affect this PR.

**Comment by kunsonx** [comment]
> @kunsonx could you update the PR with `main`? You can use the `Update branch` button here or do it manually.
> 
> There were some big changes in the codebase recently, but they should not affect this PR.

Done.

**Comment by kunsonx** [comment]
@belimawr Seems had two reviewers approved this pull request. so what is next?

**Comment by dedemorton** [comment]
I've added the backport label to make sure this gets into the next release. It sounds like this is ready to merge!

**Comment by belimawr** [comment]
> @belimawr Seems had two reviewers approved this pull request. so what is next?

Merge and wait for it to land in the next release ;)

Thanks a lot for the contribution!

**Comment by belimawr** [comment]
I was reviewing the backport and realised we missed the changelog entry :man_facepalming: 

@kunsonx would you mind creating a new PR adding the changelog entry? You can tag me to review it.

**Comment by kunsonx** [comment]
The changelog update at pull request: https://github.com/elastic/beats/pull/33714
@belimawr

**Pull request merged**
